File: Sorting Algorithms/Java/CountingSort.java
import java.util.Arrays;
public class CountingSort {

    /*
     Time O(n+k) | Space O(k)
     Stable
     Not adaptive
     */
    public static int[] CountingSort(int[] array){
        //Get he maximum element
        int max = Arrays.stream(array).max().getAsInt();

        //create count array
        int count[] = new int[max + 1];

        //populate count array
        for (int i=0; i<array.length; i++) {
            count[array[i]]++;
        }
        
        //sort array
        int array_idx = 0;
        for(int i =0; i<count.length; i++){
            while(count[i] != 0){
                array[array_idx] = i;
                count[i] -= 1;
                array_idx++;
            }
        }
        return array;
    }

        public static void main(String[] args) {
        int[] input = {5,1,4,2,8};
        System.out.println("Unsorted Array = " + Arrays.toString(input));
        CountingSort(input);
        System.out.println("Sorted Array = " + Arrays.toString(input) + " (Using Counting Sort)");
    }
}

File: Sorting Algorithms/C++/CountingSort.cpp
#include <iostream>

void printVector(std::vector<int> &vector){
    for(int i=0; i<vector.size(); i++){
        if(i==0){
            std::cout << "[";
        }
        if(i== vector.size()-1){
            std::cout << vector[i] << "]" << std::endl;
        } else {
            std::cout << vector[i] << ",";
        }
    }
}

int getMax(std::vector<int> &vector){
    int max = vector[0];
    for(int i=1; i<vector.size(); i++){
        if(vector[i] > max){
            max  = vector[i];
        }
    }
    return max;
}

std::vector<int> countingSort(std::vector<int> vector){
    int max = getMax(vector);
    
    //create count vector
    std::vector<int> count(max + 1);
    
    //populate count vector
    for(int i=0; i<vector.size(); i++){
        count[vector[i]]++;
    }
    
    //sort vector
    int vectorIdx = 0;
    for(int i=0; i<count.size(); i++){
        while(count[i] != 0){
            vector[vectorIdx] = i;
            count[i] -=1;
            vectorIdx++;
        }
    }
    return vector;
}


int main(int argc, const char * argv[]) {
    std::vector<int> testCase = {5,1,4,2,8};
    std::cout << "Input = ";
    printVector(testCase);
    
    
    std::vector<int> result = countingSort(testCase);
    std::cout << "Result = ";
    printVector(result);
    return 0;
}
